### ISSUE TITLE
Long term plans ?

### ISSUE THREAD

**Issue by petitben** [quickjs-ng/quickjs#272]
I was delighted when I learned about the initiative around Quickjs-NG (and even more when I read the names of the main contributors).

However, I now wonder about the viability of this fork in the long term. Meanwhile, Fabrice Bellard finally seems to have started contributing to the project again. I also saw that the public QuickJS github repo is from now apparently (?) open to external contributions/MR.

I'm probably going to use QuickJS as part of a long-term project and I admit I'm not sure which version to go with in terms of support/viability (I know that there is also another fork made by the OpenWebF team with specific improvements... :yum: ). At the moment the differences are not significant enough, but it could possibly change a lot in the long run (JIT or AOT support? I recently stumbled upon Ben Noordhuis' repo).

Do you intend to propose your improvements to the original project at some point? Will this fork continue to evolve separately? (this reminds me of the same kind of questions I had at the time of Io.js, the ephemeral fork of Node.js :sweat_smile: ). 

Thank you in advance for your answers and for the work accomplished so far.

**Comment by bnoordhuis** [comment]
It's too early to say. We're in contact with @chqrlie and we're amenable to converging in the future but there's a lot of logistics to work out first.

Things like JIT compilation may be forever out of scope for a converged project, even though I personally have a business case for a JS engine that sits halfway<sup>1</sup> between V8 and quickjs performance-wise, so it's likely that _a_ kind of fork will stay around for a long time.

<sup>1</sup> okay, maybe 1/10th-way

> at the time of Io.js, the ephemeral fork of Node.js

Ha, Node.js _is_ io.js - nothing ephemeral about it except the branding. :-)

**Comment by chqrlie** [comment]
Hello! we are in close contact and I speak to @Fabrice almost daily, so we are trying to avoid diverging. CI in NG is quite useful, as well as many new features added in the code, but we are working in parallel on different stuff to improve this baby...

We have some ideas for performance improvements without a JIT compilation:
- a better approach of polymorphic inline caching for property retrieval that will work with up to 2 levels of prototype chaining.  The current code does not always produce better performance and when it does, the gain might be a side effect of the longer lifetime of intermediary shapes.
- Using ropes and inline string fragments should substantially improve the string concatenation performance, especially in the pathological cases that are common in the test262 suite.
- A more efficient BigInt package should substantially increase the performance of the 64-bit computations.  String conversions can be significantly improved too.
- Regarding JIT compilation: QuickJS generates byte code directly in the parser and combs it with 2 extra passes to implement the cumbersome javascript semantics and perform basic peephole optimisations.  Producing decent executable code from the byte code without a higher representation of the whole function is not necessarily the best approach... we shall see if there is an elegant way to do it while keeping the project reasonably small.

**Comment by petitben** [comment]
Thanks for the "quick" answers. That's good to know that you are trying to avoid diverging.
Patience is a virtue I guess then. I'll close that issue.

> Ha, Node.js is io.js - nothing ephemeral about it except the branding. :-)

Yeah, sure - didn't mean to be disrespectful at all actually :-)
